### PR TITLE
Encrypting packets

### PR DESCRIPTION
**Pull request by yotamss**

### CODE REVIEW

File: Pulse.Core/PacketEncryptor.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        var generatedIV = new byte[aesIv.Length];
+        for (var i = 0; i < aesIv.Length; i++)
+        {
+            if (i < serialNumberAsBytes.Length)
+                generatedIV[i] = (byte)(aesIv[i] ^ serialNumberAsBytes[i]);
+            else
+                generatedIV[i] = aesIv[i];
+        }
+
+        return generatedIV;
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs]:
```suggestion
return aesIV
    .Zip(serialNumberAsBytes, (x, y) => x ^ y)
    .Cast<byte>()
    .Concat(aesIV[sizeof(int)..])
    .ToArray();
```

File: Pulse.Core/PacketEncryptor.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        var iv = CalculateIV(packet.SerialNumber);
+
+        using var aes = Aes.Create();
+        aes.Key = sharedKey;
+        aes.IV = iv;
+
+        using var ciphertext = new MemoryStream();
+        await using var cs = new CryptoStream(ciphertext, aes.CreateEncryptor(), CryptoStreamMode.Write);
+        await cs.WriteAsync(packet.Content);
+        await cs.DisposeAsync();
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs]: You are disposing `cs` twice

File: Pulse.Core/PacketEncryptor.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        this.aesIv = aesIv;
+        ecDiffieHellman = ECDiffieHellman.Create();
+    }
+
+    public byte[] GetPublicKey()
+    {
+        return ecDiffieHellman.PublicKey.ExportSubjectPublicKeyInfo();
+    }
+
+    public void SetSharedKey(byte[] otherPartyPublicKey)
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs]: Why are these methods `public` and the encryption/decryption methods are `internal`?
Seems like this whole type should be `internal` (and in that case, `internal` and `public` methods behave the same, so I usually mark them all as `public`)

File: Pulse.Core/PacketEncryptor.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        using var otherPartyECDH = ECDiffieHellman.Create();
+        otherPartyECDH.ImportSubjectPublicKeyInfo(otherPartyPublicKey, out _);
+        sharedKey = ecDiffieHellman.DeriveKeyMaterial(otherPartyECDH.PublicKey);
+    }
+
+    private byte[] CalculateIV(int serialNumber)
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs]: Place private helper methods below the main interface of the class

File: Pulse.Core/PacketEncryptor.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+        using var aes = Aes.Create();
+        aes.Key = sharedKey;
+        aes.IV = iv;
+
+        using var decryptedMessage = new MemoryStream(encryptedPacket.Content.ToArray());
+        await using var cs = new CryptoStream(decryptedMessage, aes.CreateDecryptor(), CryptoStreamMode.Read);
+        var contentStream = new MemoryStream();
+        await cs.CopyToAsync(contentStream);
+        await cs.DisposeAsync();  //
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs]: double disposal again

File: Pulse.Core/PacketEncryptor.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        if (sharedKey == null)
+            throw new InvalidOperationException("Shared key is not set");
+
+        var iv = CalculateIV(encryptedPacket.SerialNumber);
+
+        using var aes = Aes.Create();
+        aes.Key = sharedKey;
+        aes.IV = iv;
+
+        using var decryptedMessage = new MemoryStream(encryptedPacket.Content.ToArray());
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs]: Seems like that's the _encrypted_ message

File: Pulse.Core/PacketEncryptor.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        var decrypted = await bob.DecryptAsync(encrypted);
+        Console.WriteLine("Result:");
+        Console.WriteLine(BitConverter.ToInt32(decrypted.Content.Span));
+    }
+}
+
+
+public class PacketEncryptor : IDisposable
+{
+    private readonly byte[] aesIv;
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs]: Uppercase 'V'

**Review by GurGaller — CHANGES_REQUESTED** [review]
I know it's a draft, but I left a few technical comments anyway.
The main idea seems good (and straightforward)

File: Pulse.Core/PacketEncryptor.cs (hunk 1 above)
Comment (yotamss) [reply Pulse.Core/PacketEncryptor.cs]: nice one!

File: Pulse.Core/PacketEncryptor.cs (hunk 2 above)
Comment (yotamss) [reply Pulse.Core/PacketEncryptor.cs]: On purpose actually

File: Pulse.Core/Calls/ConnectionInfo.cs [hunk 8]
```diff
@@ -1,3 +1,9 @@
 namespace Pulse.Core.Calls;
 
-internal record ConnectionInfo(string RemoteIPAddress, int MinPort, int MaxPort);
\ No newline at end of file
+internal record ConnectionInfo(
+    string IPAddress,
+    int MinPort,
+    int MaxPort,
+    byte[] PublicKey,
+    byte[]? IV
```
Comment (GurGaller) [inline Pulse.Core/Calls/ConnectionInfo.cs]:
```suggestion
    byte[]? IV = null
```

File: Pulse.Core/Calls/ConnectionInfo.cs [hunk 9]
```diff
@@ -1,3 +1,9 @@
 namespace Pulse.Core.Calls;
 
-internal record ConnectionInfo(string RemoteIPAddress, int MinPort, int MaxPort);
\ No newline at end of file
+internal record ConnectionInfo(
+    string IPAddress,
+    int MinPort,
+    int MaxPort,
+    byte[] PublicKey,
```
Comment (GurGaller) [inline Pulse.Core/Calls/ConnectionInfo.cs:7]: Are you sure byte arrays play well with JSON deserialization?
I assume they will be serialized as JSON arrays of numbers by default, which is less efficient than passing them as ASCII strings, like I did in the server. In any case, either the server or the client has to change to conform to a common interface

File: Pulse.Core/Calls/UdpCallAcceptor.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+                    body,
+                    cancellationToken: ct
+                );
+                response.EnsureSuccessStatusCode();
+
+                return (
+                    await response.Content.ReadFromJsonAsync<ConnectionInfo>(cancellationToken: ct)
+                )!;
+            },
+            ct
```
Comment (GurGaller) [inline Pulse.Core/Calls/UdpCallAcceptor.cs:39]: Why the weird formatting?

File: Pulse.Core/Calls/UdpStreamFactory.cs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+internal class UdpStreamFactory
+{
+    public async Task<Stream> ConnectAsync(
+        Func<ConnectionInfo, Task<ConnectionInfo>> exchangeConnectionInfo,
+        CancellationToken cancellationToken = default
+    )
+    {
+        var portBruteForcer = new PortBruteForceNatTraversal();
+        var (myIPv4Address, min, max) = await portBruteForcer.PredictMinMaxPortsAsync(
+            cancellationToken
```
Comment (GurGaller) [inline Pulse.Core/Calls/UdpStreamFactory.cs:16]: ???
What's wrong with a one-line method call?

File: Pulse.Core/Connections/UdpChannel.cs [hunk 12]
```diff
@@ -26,6 +26,8 @@ private async Task ListenAsync()
 
             if (message.Buffer.All(b => b == 0) && message.Buffer.Length is 472)
             {
+                // TODO: Remove part from here...
+                // TODO: Assert that a packet length is exactly XXX (number we will decide later), this is important.
```
Comment (GurGaller) [inline Pulse.Core/Connections/UdpChannel.cs:30]: Isn't that what we did already?

File: Pulse.Core/EncryptedConnection.cs [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+{
+    private readonly IConnection actualConnection;
+    private readonly PacketEncryptor encryptor;
+    public ChannelReader<Packet> IncomingAudio { get; }
+
+    public EncryptedConnection(IConnection actualConnection, PacketEncryptor encryptor)
+    {
+        if (!encryptor.Ready)
+        {
+            throw new InvalidOperationException("Encryptor is not properly configured!");
```
Comment (GurGaller) [inline Pulse.Core/EncryptedConnection.cs]: What exactly is misconfigured?
Use a more helpful error message

File: Pulse.Core/EncryptedConnection.cs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+    private class CryptoPacketChannelReader : ChannelReader<Packet>
+    {
+        private readonly ChannelReader<Packet> actualChannelReader;
+        private readonly PacketEncryptor encryptor;
+
+        public CryptoPacketChannelReader(
+            ChannelReader<Packet> actualChannelReader,
+            PacketEncryptor encryptor
+        )
```
Comment (GurGaller) [inline Pulse.Core/EncryptedConnection.cs:39]: The accepted convention is to place the ending parenthesis of a method next to the last parameter, and not on a separate line (that also holds for method calls)

File: Pulse.Core/EncryptedConnection.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            if (!actualChannelReader.TryRead(out item))
+                return false;
+
+            item = encryptor.DecryptAsync(item).GetAwaiter().GetResult(); // Todo make the encrypt and decrypt methods sync
+            return true;
+        }
+
+        public override ValueTask<bool> WaitToReadAsync(
+            CancellationToken cancellationToken = new CancellationToken()
```
Comment (GurGaller) [inline Pulse.Core/EncryptedConnection.cs]:
```suggestion
            CancellationToken cancellationToken = default)
```

File: Pulse.Core/PacketEncryptor.cs [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    public byte[] PublicKey => ecDiffieHellman.PublicKey.ExportSubjectPublicKeyInfo();
+
+    public void SetOtherPartyPublicKey(byte[] otherPartyPublicKey)
+    {
+        using var otherPartyECDH = ECDiffieHellman.Create();
+        otherPartyECDH.ImportSubjectPublicKeyInfo(otherPartyPublicKey, out _);
+        sharedKey = ecDiffieHellman.DeriveKeyMaterial(otherPartyECDH.PublicKey);
+    }
+
+    public void SetAesIv(byte[] iv)
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs]:
```suggestion
    public void SetAesIV(byte[] iv)
```

File: Pulse.Core/PacketEncryptor.cs [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        aesIV = iv;
+    }
+
+    public bool Ready => sharedKey != null && aesIV != null;
+
+    public async Task<Packet> EncryptAsync(Packet packet)
+    {
+        if (!Ready)
+            throw new InvalidOperationException("Packet encryptor is not sufficiently configured.");
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs:44]: An unhelpful error message

File: Pulse.Core/PacketEncryptor.cs [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+        otherPartyECDH.ImportSubjectPublicKeyInfo(otherPartyPublicKey, out _);
+        sharedKey = ecDiffieHellman.DeriveKeyMaterial(otherPartyECDH.PublicKey);
+    }
+
+    public void SetAesIv(byte[] iv)
+    {
+        aesIV = iv;
+    }
+
+    public bool Ready => sharedKey != null && aesIV != null;
```
Comment (GurGaller) [inline Pulse.Core/PacketEncryptor.cs]:
```suggestion
    [MemberNotNullWhen(true, nameof(sharedKey)]
    [MemberNotNullWhen(true, nameof(aesIV)]
    public bool Ready => sharedKey is not null && aesIV is not null;
```

**Review by GurGaller — CHANGES_REQUESTED** [review]
I don't know if it's intentional, but something is off about the formatting in this PR (see comments)

File: Pulse.Core/Connections/UdpChannel.cs (hunk 12 above)
Comment (yotamss) [reply Pulse.Core/Connections/UdpChannel.cs:30]: It should not be here, packets are encrypted..

File: Pulse.Core/Calls/ConnectionInfo.cs (hunk 8 above)
Comment (yotamss) [reply Pulse.Core/Calls/ConnectionInfo.cs]: We should separate it into two different records.

File: Pulse.Core/Calls/UdpCallAcceptor.cs (hunk 10 above)
Comment (yotamss) [reply Pulse.Core/Calls/UdpCallAcceptor.cs:39]: CSharper, a fucked up plugin. please reformat the whole file

File: Pulse.Core/EncryptedConnection.cs (hunk 13 above)
Comment (yotamss) [reply Pulse.Core/EncryptedConnection.cs]: There are only two options so I left it like that for readability, using `Ready` is much more readable.

File: Pulse.Core/PacketEncryptor.cs (hunk 18 above)
Comment (yotamss) [reply Pulse.Core/PacketEncryptor.cs]: nice.

**Pull request merged**